### PR TITLE
Initial replace of circleci to use github actions

### PR DESCRIPTION
**Pull request by cpanato**
#### Summary
Initial work to replace CircleCI to use GitHub Actions.
In this PR we are moving the Checks jobs and golangci-lint

#### Ticket Link
n/a

#### Release Note

```release-note
NONE
```

### CODE REVIEW

**Comment by cpanato** [comment]
@agnivade looks like there are some lints to fix and the `check-migrations` is failing, is it okay for you that I update those?

**Comment by cpanato** [comment]
tagging @isacikgoz @lieut-data @streamer45 for visibility, tag others if needed

**Comment by cpanato** [comment]
@amyblais this should be cherry-picked to the active release branches

**Comment by amyblais** [comment]
> @amyblais this should be cherry-picked to the active release branches when approved and merged

The current active branches are: ``release-6.3``, ``release-6.2``, ``release-6.1``, ``release-5.37``, and ``cloud``.

**Comment by agnivade** [comment]
@cpanato - It seems like there is no way to set up dependent steps like we did in CircleCI is there? This seems to be just running all jobs in parallel.

File: .github/workflows/checks.yml [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+      with:
+        go-version: ${{ env.GOVERSION }}
+
+    - name: Run check-mocks
+      run: |
+        make einterfaces-mocks store-mocks
+        if [[ -n $(git status --porcelain) ]]; then echo "Please update the mocks using 'make einterfaces-mocks store-mocks'"; exit 1; fi
+
+  check-migrations:
+    runs-on: ubuntu-latest
```
Comment (agnivade) [inline .github/workflows/checks.yml:76]: This does not seem to run on the mattermost docker image we have.

File: .github/workflows/checks.yml [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+      with:
+        go-version: ${{ env.GOVERSION }}
+
+    - name: Run check-gen-serialized
+      run: |
+        make gen-serialized
+        if [[ -n $(git status --porcelain) ]]; then echo "Please update the serialized files using 'make gen-serialized'"; exit 1; fi
+
+  check-mattermost-vet:
+    runs-on: ubuntu-latest
```
Comment (agnivade) [inline .github/workflows/checks.yml:140]: same here as well.

**Review by agnivade — COMMENTED** [review]
@cpanato - I'd like to understand this change very well since this is the main repo. Please give me some time to go through the changes thoroughly.

cc @jupenur as well.

File: .github/workflows/checks.yml (hunk 2 above)
Comment (cpanato) [reply .github/workflows/checks.yml:140]: for those we just need go to be available, there is not extra dependency

**Comment by cpanato** [comment]
> @cpanato - I'd like to understand this change very well since this is the main repo. Please give me some time to go through the changes thoroughly.
> 
> cc @jupenur as well.

sure take your time, this is the initial migration from circleci to use github actions, that I mention to you in our call and also gave a psa in the dev meeting this week

**Comment by cpanato** [comment]
> @cpanato - It seems like there is no way to set up dependent steps like we did in CircleCI is there? This seems to be just running all jobs in parallel.

those check were running in parallel before (in circleci)

File: .github/workflows/checks.yml (hunk 1 above)
Comment (cpanato) [reply .github/workflows/checks.yml:76]: in circleci there is no easy way to install go or have that available in the env, that is why we run over a container.
we can do that here if you want as well

**Comment by agnivade** [comment]
> sure take your time, this is the initial migration from circleci to use github actions, that I mention to you in our call and also gave a psa in the dev meeting this week

Yep, I would like to understand this well before merging anything.

> those check were running in parallel before (in circleci)

Yes, I understand "some" checks were in parallel. But as you can see, they were dependent on `setup`, and the following stages are dependent on those. Is that possible here?

**Comment by cpanato** [comment]
> > sure take your time, this is the initial migration from circleci to use github actions, that I mention to you in our call and also gave a psa in the dev meeting this week
> 
> Yep, I would like to understand this well before merging anything.
> 
> > those check were running in parallel before (in circleci)
> 
> Yes, I understand "some" checks were in parallel. But as you can see, they were dependent on `setup`, and the following stages are dependent on those. Is that possible here?

the setup is needed for the testing / build part, not for the checks because the setup fetches the Webapp repository, only if the those checks needs the Webapp I understood wrong the checks and will refactor, otherwise it is not needed

**Comment by agnivade** [comment]
This is needed for early exit. If setup fails, no other steps will run. Otherwise, every time all stages will run and increase overall CI execution time even if we know all other steps will fail.

**Comment by cpanato** [comment]
> This is needed for early exit. If setup fails, no other steps will run. Otherwise, every time all stages will run and increase overall CI execution time even if we know all other steps will fail.

setup just run the checkout there is not other checks there

**Comment by cpanato** [comment]
thanks for your review and feedback @agnivade based on our conversation over zoom I've made some changes

also, when we migrate the tests/build they will depends on the checks

**Comment by cpanato** [comment]
closing per @spirosoik request

**Pull request closed without merging**